src/commands/recover.py

```diff
@@ -0,0 +1,232 @@
+    )
+
+    try:
+        asyncio.run(
+            main(
+                mnemonic,
+                keystores_dir,
+                password_file,
+                per_keystore_password,
+                config,
+            )
+        )
+    except Exception as e:
+        log_verbose(e)
+
+
+# pylint: disable-next=too-many-arguments
+async def main(
+    mnemonic: str,
+    keystores_dir: Path,
+    password_file: Path,
+    per_keystore_password: bool,
+    config: VaultConfig,
+):
+    validators = await _fetch_registered_validators()
+    if not validators:
+        raise click.ClickException('Registered validators not found')
+    click.secho(f'Found {len(validators)} validators, start recovering...')
+
+    mnemonic_next_index = await _generate_keystores(
+        mnemonic,
+        keystores_dir,
+        password_file,
+        validators,
+        per_keystore_password,
+    )
+
+    config.save(settings.network, mnemonic, mnemonic_next_index)
+    click.secho(f'Vault {settings.vault} successfully recovered', bold=True, fg='green')
+
+
+# pylint: disable-next=too-many-locals
+async def _fetch_registered_validators() -> list[RegisteredValidator]:
+    """Fetch registered validators."""
+    block = await execution_client.eth.get_block('latest')
+    current_block = block['number']
+    keeper_genesis_block = settings.network_config.KEEPER_GENESIS_BLOCK
+
+    page_size = 50_000
+    public_keys = []
+
+    for cursor in range(keeper_genesis_block, current_block, page_size):
+        page_start = cursor
+        page_end = min(cursor + page_size - 1, current_block)
+
+        events = await vault_contract.events.ValidatorRegistered.get_logs(
+            fromBlock=page_start, toBlock=page_end
+        )
+        for event in events:
+            hex_key = event['args']['publicKey'].hex()
+            public_keys.append(add_0x_prefix(hex_key))
+
+    results = []
+
+    for i in range(0, len(public_keys), settings.validators_fetch_chunk_size):
+        validators = await consensus_client.get_validators_by_ids(
+            public_keys[i : i + settings.validators_fetch_chunk_size]
+        )
+        for beacon_validator in validators['data']:
+            results.append(
+                RegisteredValidator(
+                    index=beacon_validator['index'],
+                    public_key=beacon_validator['validator']['pubkey'],
+                    status=beacon_validator['status'],
+                )
+            )
+
+    return results
+
+
+# pylint: disable-next=too-many-arguments,too-many-locals
+async def _generate_keystores(
+    mnemonic: str,
+    keystores_dir: Path,
+    password_file: Path,
+    validators: list[RegisteredValidator],
+    per_keystore_password: bool,
+):
+    keystores_dir.mkdir(parents=True, exist_ok=True)
+    exited_statuses = [x.value for x in EXITING_STATUSES]
+    index = 0
+    failed_attempts = 0
+
+    while not all(v.keystore_found for v in validators):
```

Comment:
Currently, there are too many iterations over validators list. I think it's better to use counter instead of all(...). Also, instead of checking every validator, you can have a dict[pub_key, status] (checking against dict key is much faster).